tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd

```diff
@@ -147,131 +164,79 @@ Note that:
 
 #### Step 3: Create the pPCA model
 
-```julia
-@model function pPCA(x, ::Type{TV}=Array{Float64}) where {TV}
-    # Dimensionality of the problem.
-    N, D = size(x)
-
-    # latent variable z
-    z ~ filldist(Normal(), D, N)
-
-    # side note for the curious
-    # we use the more concise filldist syntax partly for compactness, but also for compatibility with other AD
-    # backends, see the [Turing Performance Tipps](https://turing.ml/dev/docs/using-turing/performancetips)
-    # w = TV{2}(undef, D, D)
-    # for d in 1:D
-    #  w[d, :] ~ MvNormal(ones(D))
-    # end
-
-    # weights/loadings W
-    w ~ filldist(Normal(), D, D)
-
-    # mean offset
-    m ~ MvNormal(ones(D))
-    mu = (w * z .+ m)'
-    for d in 1:D
-        x[:, d] ~ MvNormal(mu[:, d], ones(N))
-    end
-end;
-```
+Here we construct the probabilistic model `pPCA()`. As per the p-PCA formula, we think of each row (i.e. each gene feature) following a $N=60$ dimensional multivariate normal distribution centered around the corresponding row of $\mathbf{W}_{D \times k} \times \mathbf{Z}_{k \times N} + \boldsymbol{\mu}_{D \times N}$.
 
 ```julia
-@model function pPCA(x, k)
-    # function pPCA(k::Int64, x::Type{TV}=Array{Float64}) where {TV}
+@model function pPCA(X::Array{Float64}, k::Int64)
     # retrieve the dimension of input matrix x.
-    N, D = size(x)
+    # N, D = size(x)
+    D, N = size(X)
 
+    # weights/loadings/projection matrix W
+    W ~ filldist(Normal(), D, k)
     # latent variable z
-    z ~ filldist(Normal(), k, N)
-
-    # side note for the curious
-    # we use the more concise filldist syntax partly for compactness, but also for compatibility with other AD
-    # backends, see the [Turing Performance Tipps](https://turing.ml/dev/docs/using-turing/performancetips)
-    # w = TV{2}(undef, D, D)
-    # for d in 1:D
-    #  w[d, :] ~ MvNormal(ones(D))
-    # end
-
-    # weights/loadings W
-    w ~ filldist(Normal(), D, k)
-
-    # mean offset
-    m ~ MvNormal(ones(D))
-    for d in 1:D
-        mu_d = ((w * z)[d, :] .+ m[d]) # mu_d is an N x 1 vector representing the d-th feature values for all instances.
-        x[:, d] ~ MvNormal(mu_d, ones(N))
+    Z ~ filldist(Normal(), k, N)
+
+    # uniform offset across features
+    μ ~ MvNormal(ones(N))
+    # likelihood
+    for gene_ind in 1:D
+        gene_mean = (W * Z)[gene_ind, :] .+ μ
+        X[gene_ind, :] ~ MvNormal(gene_mean, ones(N))
     end
```

Comment:
Again this uses deprecated syntax for `MvNormal`. Moreover, recomputing `W*Z` in every step is very inefficient (and the row-wise access is slow). Use something like
```suggestion
    genes_mean = Z' * W' .+ μ
    Xt = X'
    Xt ~ arraydist([MvNormal(m, Eye(N)) for m in eachcol(genes_mean)])
```
(the transposes indicate that `X` should rather be provided with dimensions reversed and `W` and `Z` should also be defined differently)

Reply:
Yes, moving `W * Z` out of the loop could save computational efforts (e.g. defining `WZ = W * Z` before the `for` loop). Also thanks for suggesting using `arraydist` - that's even better, it avoids using `for` loop, as suggested in the Performance Tips by `Turing.jl` (https://turing.ml/dev/docs/using-turing/performancetips).
So I re-write the model following the suggestions.